[Sreenivasa Kalpana Surya] iP

Tetrerox opened the pull request:
# UhtredRagnarson
> Courage is finding the will to overcome your fear, nothing more.
>  -Uhtred Ragnarson

#### UhtredRagnarson remembers things for you, so *you* don't have to. It's 
* text-based
* easy to use
* has ~~some~~ **NO** learning curve

#### All you need to do is, 
1. download it from [here](https://github.com/Tetrerox/ip/releases/tag/A-Jar)
2. open it without *any* installation
3. Add your tasks
4. And watch the **magic** happen! 👍 

And the *best* part is that you can use it even if you are **BROKE!**

#### Features:
- [X] Add tasks
- [X] Delete tasks
- [X] Mark tasks as completed
- [X] Find matching tasks based on a keyword
- [X] GUI
- [ ] Custom replies to *each* command
- [ ] Better UI design

If you **LOVE** java programming, then you can learn from this project too! Here's the main method:

```java
public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
```

Commit:
Add Gradle support

Tetrerox committed:
Completed Level-1

Tetrerox committed:
Completed Level-2

Tetrerox committed:
Completed Level-3

Tetrerox committed:
Completed Level-4

Tetrerox committed:
Completed A-TextUiTesting and renamed my bot

Tetrerox committed:
Completed Level-5

Tetrerox committed:
Completed Level-6

Tetrerox committed:
Add the ability to read and write files and load from pre-existing data

Tetrerox committed:
Deadline can now accept dates in the format yyyy-mm-dd and parse them into text and print them into dd mmm yyyy

Tetrerox committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Lawbringer.java

Tetrerox committed:
Add Storage.java and abstract it from Lawbringer.java

Tetrerox committed:
Add Parser.java and Ui.java and abstract it from Lawbringer.java

Tetrerox committed:
Add TaskList.java

Tetrerox committed:
Organize the classes into packages

Tetrerox committed:
Add JUnit tests

Tetrerox committed:
Revert "Add JUnit tests"

This reverts commit 896b7a6f983a250ba6c3ce6f69c1bea13b1f9a03.

Tetrerox committed:
no message

Tetrerox committed:
Create jar file

Tetrerox committed:
Add javadoc comments and adjust the coding style to adhere to the coding standard.

Tetrerox committed:
Give users a way to find a task by searching for a keyword.

Tetrerox committed:
Remove lib folder from version control

Tetrerox committed:
Update .gitignore file to ignore lib folder

Tetrerox committed:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d' into branch-A-Gradle

Tetrerox committed:
Fix checkstyle errors

Tetrerox committed:
Rename the chatbot

Tetrerox committed:
Add GUI but incomplete. Need to fix file storage.

elroygohjy commented on src/main/java/lawbringer/Storage.java:
```diff
@@ -0,0 +1,91 @@
+        Scanner s = new Scanner(file);
+        while (s.hasNext()) {
+            String line = s.nextLine();
+            char taskType = line.charAt(1);
+            boolean isDone = line.charAt(4) == 'X';
+            switch (taskType) {
+            case 'T':
+                ToDo todo = new ToDo(line.substring(7));
+                if (isDone) {
+                    todo.markAsDone();
+                }
+                tasks.add(todo);
+                break;
+            case 'D':
+                int byIndex = line.indexOf(" (by: ");
+                Deadline deadline = new Deadline(line.substring(7, byIndex),
+                        line.substring(byIndex + 6, line.length() - 1));
+                if (isDone) {
+                    deadline.markAsDone();
+                }
+                tasks.add(deadline);
+                break;
+            case 'E':
+                int atIndex = line.indexOf(" (at: ");
+                Event event = new Event(line.substring(7, atIndex),
+                        line.substring(atIndex + 6, line.length() - 1));
```
Should the constants be extracted out as final constant variables?

elroygohjy commented on src/main/java/lawbringer/TaskList.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    /**
+     * Add a deadline task to the list and prints a message to the user.
+     *
+     * @param userInput The input of the user.
+     * @param ui        Ui class to print messages to the user.
+     */
+    protected void addDeadlineTask(String userInput, Ui ui) throws DateTimeParseException {
```
Should you include @ throws as you are throwing DateTimeParseException?

elroygohjy commented on src/main/java/lawbringer/ToDo.java:
```diff
@@ -0,0 +1,21 @@
+     *
+     * @param title Description of the to-do.
+     */
+    public ToDo(String title) {
+        super(title);
```
Do you think title as a name for description is abit ambiguous?

elroygohjy reviewed (commented):
I think you have done a great job! The testing part is very clean.

Tetrerox committed:
Add GUI

Tetrerox committed:
Remove .txt file which was added accidentaly

Tetrerox committed:
Add GUI

zhenxuantan commented on src/main/java/uhtredragnarson/Deadline.java:
```diff
@@ -0,0 +1,26 @@
+ */
+public class Deadline extends Task {
+
+    protected String by;
```
While short, I didn't like the naming for the variable "by". Shouldn't both Deadline and Event have the same variable name (and also they do not consist of only timing) even though the tags "/at" and "/by" are different?

zhenxuantan commented on src/main/java/uhtredragnarson/UhtredRagnarson.java:
```diff
@@ -0,0 +1,47 @@
+ */
+public class UhtredRagnarson {
+
+    private final Storage storage;
```
Shouldn't these variables be in uppercase? I think the IntelliJ detects it as a warning but we probably have to follow the convention! 😄

zhenxuantan reviewed (commented):
Overall the code is very readable and very clean to read! 😄

Tetrerox committed:
There are assumptions in the code

Making assumptions in the code can lead to possible bugs and the sooner the execution stops, the safer it is.

When the .txt file that the code is trying to read from is missing, we assert that a new directory and file has indeed been created.

Adding these assertions will make sure that the program will stop abruptly when a new directory and file has not been created as they will produce errors when we try to modify the file that does not exist.

Tetrerox committed:
Code has commented out code and a pre-initialised variable

Old code is redundant and pre-initialising a variable and not using that value serves no purpose.

Old code is removed and the variable is now declared but not initialised.

Old code can always be recovered from version control if you need to use it again.

Tetrerox committed:
Merge pull request #2 from Tetrerox/branch-A-Assertions

There are assumptions in the code

Making assumptions in the code can lead to possible bugs and the sooner the execution stops, the safer it is.

When the .txt file that the code is trying to read from is missing, we assert that a new directory and file has indeed been created.

Adding these assertions will make sure that the program will stop abruptly when a new directory and file has not been created as they will produce errors when we try to modify the file that does not exist.

Tetrerox committed:
Merge branch 'master' into branch-A-CodeQuality

Tetrerox committed:
Merge pull request #1 from Tetrerox/branch-A-CodeQuality

Code has commented out code and a pre-initialised variable

Old code is redundant and pre-initialising a variable and not using that value serves no purpose.

Old code is removed and the variable is now declared but not initialised.

Old code can always be recovered from version control if you need to use it again.

Tetrerox committed:
Add a command to access a help page (C-Help)

Tetrerox committed:
Add time to deadline and fix a storage bug

Tetrerox committed:
Add new date and time format to event

Tetrerox committed:
Add abstraction to TaskList.java

Tetrerox committed:
Add better GUI

Tetrerox committed:
Repackage files

Tetrerox committed:
Add more JUnit tests

Tetrerox committed:
Add Ui.png

Tetrerox committed:
Add UserGuide

Tetrerox committed:
Edit UserGuide

Tetrerox committed:
Edit UserGuide

Tetrerox committed:
Update README.md

Tetrerox committed:
Release JAR file

Tetrerox committed:
Handle more exceptions

Tetrerox committed:
Edit jar file name

Tetrerox committed:
Ensure that .jar file works on other PCs
